### ISSUE TITLE
Filters on ids prevent sorting and limit

### ISSUE THREAD

**Issue by To-om** [JanusGraph/janusgraph#3047]
- JanusGraph 0.6.1

When vertex IDs are filtered using `has`, the limit, range and order steps don't work:
```
gremlin> graph = JanusGraphFactory.build().set('storage.backend', 'inmemory').open()
gremlin> GraphOfTheGodsFactory.loadWithoutMixedIndex(graph, true)
gremlin> g = graph.traversal()
gremlin> g.V().has("age").id().fold()
==>[4168,8264,12360,16456,8360,12456]

gremlin> g.V().has(T.id, P.within(4168,8264,12360,16456,8360,12456)).limit(2)
==>v[4168]
==>v[8264]
==>v[12360]
==>v[16456]
==>v[8360]
==>v[12456]

gremlin> g.V().hasId(4168).limit(0)
==>v[4168]

gremlin> g.V().has(T.id, P.within(4168,8264,12360,16456,8360,12456)).order().by("age").values("age")
==>10000
==>4500
==>30
==>4000
==>5000
==>45
```
Explain of the query:
```
g.traversal().V().hasId(4096).limit(0).explain()
==>Traversal Explanation
============================================================================================================================
Original Traversal                                     [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]

ConnectiveStrategy                               [D]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
IdentityRemovalStrategy                          [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
EarlyLimitStrategy                               [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
MatchPredicateStrategy                           [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
FilterRankingStrategy                            [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
InlineFilterStrategy                             [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
IncidentToAdjacentStrategy                       [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
AdjacentToIncidentStrategy                       [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
ByModulatorOptimizationStrategy                  [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
RepeatUnrollStrategy                             [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
CountStrategy                                    [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
PathRetractionStrategy                           [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
LazyBarrierStrategy                              [O]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
AdjacentVertexFilterOptimizerStrategy            [P]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
AdjacentVertexHasIdOptimizerStrategy             [P]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
AdjacentVertexIsOptimizerStrategy                [P]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
AdjacentVertexHasUniquePropertyOptimizerStrategy [P]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
JanusGraphLocalQueryOptimizerStrategy            [P]   [GraphStep(vertex,[]), HasStep([~id.eq(4096)]), RangeGlobalStep(0,0)]
JanusGraphStepStrategy                           [P]   [JanusGraphStep(vertex,[4096])]
JanusGraphMultiQueryStrategy                     [P]   [JanusGraphStep(vertex,[4096])]
JanusGraphMixedIndexCountStrategy                [P]   [JanusGraphStep(vertex,[4096])]
JanusGraphIoRegistrationStrategy                 [P]   [JanusGraphStep(vertex,[4096])]
ProfileStrategy                                  [F]   [JanusGraphStep(vertex,[4096])]
StandardVerificationStrategy                     [V]   [JanusGraphStep(vertex,[4096])]

Final Traversal                                        [JanusGraphStep(vertex,[4096])]
```
The RangeGlobalStep was removed in the final traversal.

The range and order steps work if IDs are in V(): `g.V(4096,8192,12288,16384,20480).limit(2).count()` returns 2

Moreover, I notice that the index backend is not used when IDs are provided. I think the performance of queries with high number of IDs that have extra filters (using full-text for example) and sorting could be improved if index is used.

**Comment by li-boxuan** [comment]
I can confirm this bug in v0.5 and v0.6 branches.

**Comment by To-om** [comment]
A simplistic fix could be:
```patch
diff --git a/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphStepStrategy.java b/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphStepStrategy.java
index bf3cc2df0..828a5bde8 100644
--- a/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphStepStrategy.java
+++ b/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphStepStrategy.java
@@ -48,8 +48,10 @@ public class JanusGraphStepStrategy extends AbstractTraversalStrategy<TraversalS
                 TraversalHelper.replaceStep(originalGraphStep, janusGraphStep, traversal);
                 HasStepFolder.foldInIds(janusGraphStep, traversal);
                 HasStepFolder.foldInHasContainer(janusGraphStep, traversal, traversal);
-                HasStepFolder.foldInOrder(janusGraphStep, janusGraphStep.getNextStep(), traversal, traversal, janusGraphStep.returnsVertex(), null);
-                HasStepFolder.foldInRange(janusGraphStep, JanusGraphTraversalUtil.getNextNonIdentityStep(janusGraphStep), traversal, null);
+                if (janusGraphStep.getIds().length == 0) {
+                    HasStepFolder.foldInOrder(janusGraphStep, janusGraphStep.getNextStep(), traversal, traversal, janusGraphStep.returnsVertex(), null);
+                    HasStepFolder.foldInRange(janusGraphStep, JanusGraphTraversalUtil.getNextNonIdentityStep(janusGraphStep), traversal, null);
+                }
             } else {
                 //Make sure that any provided "start" elements are instantiated in the current transaction
                 final Object[] ids = originalGraphStep.getIds();
```

**Comment by li-boxuan** [comment]
@To-om Thanks for the fix suggestion! That should work, but I would prefer fixing it in JanusGraphStep.java for code readability and logic consistency. If you are interested in submitting a patch, that would be great! Otherwise I can fix it in a day or two.

**Comment by porunov** [comment]
Fixed in #3055
